### PR TITLE
Feat: Refactor name parsing to its own class and add namespace parsing

### PR DESCRIPTION
**Pull request by Siphonophora**
Refactoring to make it easier to share across loggers. Added namespace parsing and tests because the nunit logger needed it.

### CODE REVIEW

File: src/JUnit.Xml.TestLogger/TestCaseNameParser.cs [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        public static ParsedName Parse(string fullyQualifedName)
+        {
+            var metadataNamespaceName = string.Empty;
+            var metadataTypeName = string.Empty;
+            var metadataMethodName = string.Empty;
+
+            var step = NameParseStep.FindMethod;
+            var state = NameParseState.Default;
+
+            var tuptuo = new List<char>();
```
Comment (codito) [inline src/JUnit.Xml.TestLogger/TestCaseNameParser.cs]: suggestion: may be use a better name for the variable?

File: src/JUnit.Xml.TestLogger/TestCaseNameParser.cs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+            try
+            {
+                var eman = fullyQualifedName.ToCharArray().Reverse().ToArray();
```
Comment (codito) [inline src/JUnit.Xml.TestLogger/TestCaseNameParser.cs]: Suggest to use the index accessor in string directly to save a new array allocation and traversal.

`for (int i = fullyQualifiedName.Length-1; i >= 0; i--)` followed by `fullyQualifiedName[i]`

File: src/JUnit.Xml.TestLogger/TestCaseNameParser.cs [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+                        }
+                    }
+                    else
+                    {
+                        // We are in String State.
+                        if (thisChar == '"')
+                        {
+                            if (eman.ElementAtOrDefault(i + 1) == '\\')
+                            {
+                                // The quote was escaped, so its atually a quote mark in a string
```
Comment (codito) [inline src/JUnit.Xml.TestLogger/TestCaseNameParser.cs]: `s/atually/actually`

File: src/JUnit.Xml.TestLogger/TestCaseNameParser.cs [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+                                tuptuo.Add(thisChar);
+                            }
+                            else
```
Comment (codito) [inline src/JUnit.Xml.TestLogger/TestCaseNameParser.cs]: else is redundant, the `if` can be inverted and `tuptuo.Add` may be outside `if`.

File: src/JUnit.Xml.TestLogger/TestCaseNameParser.cs [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+                    tuptuo.Reverse();
+                    metadataNamespaceName = string.Join(string.Empty, tuptuo);
+                }
+                else if (step == NameParseStep.FindType)
+                {
+                    tuptuo.Reverse();
+                    metadataTypeName = string.Join(string.Empty, tuptuo);
+                }
+            }
+            catch (Exception)
```
Comment (codito) [inline src/JUnit.Xml.TestLogger/TestCaseNameParser.cs:171]: Should we log the error?

File: src/JUnit.Xml.TestLogger/TestCaseNameParser.cs [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+                            tuptuo.Add(thisChar);
+                        }
+                        else if (thisChar == '"')
+                        {
+                            // This must come at the end of a string, when escape characters aren't an issue, so we are
+                            // 'entering' string state, because of the reverse parsing.
+                            state = NameParseState.String;
+                            tuptuo.Add(thisChar);
+                        }
+                        else
```
Comment (codito) [inline src/JUnit.Xml.TestLogger/TestCaseNameParser.cs]: `else` looks redundant.

File: src/JUnit.Xml.TestLogger/TestCaseNameParser.cs (hunk 5 above)
Comment (Siphonophora) [reply src/JUnit.Xml.TestLogger/TestCaseNameParser.cs:171]: I don't think we need to. The console write line below informs the user of the issue. If they want to report it, the error message has the FullyQualifiedName included, which is everything we would need to replicate.

File: src/JUnit.Xml.TestLogger/TestCaseNameParser.cs (hunk 6 above)
Comment (Siphonophora) [reply src/JUnit.Xml.TestLogger/TestCaseNameParser.cs]: Agreed. Fixed

File: src/JUnit.Xml.TestLogger/TestCaseNameParser.cs (hunk 1 above)
Comment (Siphonophora) [reply src/JUnit.Xml.TestLogger/TestCaseNameParser.cs]: Done. got rid of the reversed list.

File: src/JUnit.Xml.TestLogger/TestCaseNameParser.cs (hunk 2 above)
Comment (Siphonophora) [reply src/JUnit.Xml.TestLogger/TestCaseNameParser.cs]: Good idea. Done

**Review by codito — APPROVED** [review]

**Pull request merged**